### PR TITLE
Read Gmsh ASCII and Binary 2.0 files

### PR DESCRIPTION
**Pull request by martemyev**
Dear All,

Here is a first attempt to include Gmsh native mesh files into MFEM. This code works for me - I have already run some simulations. I can write some unit tests if you have an example of testing other mesh readers.

Thanks,
Mikhail

### CODE REVIEW

**Comment by tzanio** [comment]
Thank you Mikhail, this will be really great to have in MFEM!

We will try to review it this week and get back to you.

Do you mind sharing some of the meshes/geometry files you've used? How difficult it will be to also add support for Gmsh's high-order meshes?

Cheers,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,

I fixed one bug in the initial commit - should be pulled in by Github automatically.
As for the meshes - I checked with most of grids from Gmsh tutorials (they are included in the distributed Gmsh archive). You can generate the meshes on you side with:
```Bash
$ gmsh t13.geo -3
```
for example, for 13-th tutorial.
I checked the reading process of the Gmsh grids with the following simple code that initializes a Mesh object, saves it in the native MFEM format, refines the initial grid and saves it again.
```cxx
#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
   // Name of the mesh file without extension
   const char *meshfile_base = "base";

   OptionsParser args(argc, argv);
   args.AddOption(&meshfile_base, "-base", "--meshfile-base", "Name of Gmsh mesh file w/o extension");
   args.Parse();
   if (!args.Good())
   {
      args.PrintUsage(cout);
      return 1;
   }
   args.PrintOptions(cout);

   string filename = string(meshfile_base) + ".msh"; // Native extension of Gmsh meshes

   ifstream imesh(filename.c_str());
   if (!imesh)
   {
      cerr << "\nCan not open mesh file: " << filename << '\n' << endl;
      return 2;
   }

   cout << "Reading mesh from file " << filename << " ..." << flush;
   Mesh mesh(imesh, 1, 1);
   cout << "completed\n" << flush;

   imesh.close();
   const int dim = mesh.Dimension();

   // Write the mesh to GlVis format to see how it works
   {
      filename = string(meshfile_base) + ".mesh";
      ofstream mesh_ofs(filename.c_str());
      mesh_ofs.precision(8);
      mesh.Print(mesh_ofs);
   }

   // Refine the mesh to increase the resolution.
   {
      int ref_levels =
         floor(log(500000./mesh.GetNE())/log(2.)/dim);
      for (int l = 0; l < ref_levels; l++)
         mesh.UniformRefinement();
   }

   // Write the refined mesh to GlVis format
   {
      filename = string(meshfile_base) + "_refined.mesh";
      ofstream mesh_ofs(filename.c_str());
      mesh_ofs.precision(8);
      mesh.Print(mesh_ofs);
   }
}
```
I run it for the t13.msh with:
```Bash
$ ./read_gmsh -base t13
```
That produces 2 files: t13.mesh and t13_refined.mesh. To check if the mesh was initialized correctly with the Gmsh grid, I open these 2 files with GlVis. For this t13 example, everything works fine. However, for many other tutorial grids, it doesn't. That's mostly because Gmsh generates some of grids with 0 physical ID that is used as an attribute, and therefore gives an error on MFEM side. This is okay, I believe. The only thing that bothers me is that for some simple 2D grids like from tutorial 1 (t1.geo), I have the following view in GlVis

t1

Why are there 3 layers of triangles? The t1.mesh file looks good, though. The same happens with tutorials 6 and 7.

Also, I couldn't figure out why the following happens with tutorial 3 that has a quite fancy grid with different types of cells (including prisms that are not supported by MFEM, and therefore just skipped when the mesh is initialized)
t3

Basically, I tried to verify the Gmsh reading procedure by visual inspection and comparison with GLVis. If you have a more rigorous approach, I'll be happy to use it.

As for the higher order meshes - I didn't think about that yet. That would be very nice to have too.

Best,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

Thanks for the explanations.

We also discovered and fixed the bug with `elements_0D`.

Our current fix for the 0 element attribute is to replace it by 1. Alternatively, we can call `Mesh::SetAttributes()` and generalize it to shift the attributes to be positive. This is more general, but it may change the Gmsh attribute numbering more dramatically. Which approach will you prefer?

What you see in GLVis is the piecewise-constant function that represents a coloring of the mesh. In this case, 3 colors are needed so face neighbors have different colors. If you open `data/inline-quad.mesh`, you should see two layers.

The issue with tutorial 3 is the boundary (probably from the prismatic elements). If I delete the boundary elements in the *.mesh file (i.e. "boundary\n0"), I can see the tet portion of the mesh.

Out of the other tutorials, `t9` is not really a mesh; `t12` doesn't work because it is a surface mesh; and `t2`, `t6` and `t15` break because they combine meshes of different dimensions. The rest seem to work fine for me.

Cheers,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,

> Our current fix for the 0 element attribute is to replace it by 1. Alternatively, we can call Mesh::SetAttributes() and generalize it to shift the attributes to be positive. This is more general, but it may change the Gmsh attribute numbering more dramatically. Which approach will you prefer?

In fact, I would go a third way and make it impossible to initialize an element with a non-positive attribute when reading a Gmsh file. I mean, right now an MFEM Warning is showed about that, but I'd throw an error. That's because in Gmsh it's discouraged (though possible) to save an element with non-positive physical ID (which is used as an attribute in MFEM), and 0 is reserved for internal Gmsh numbering, for example, when a user didn't specify any elements for saving, and just saved everything. In this case the elements will have 0 ID. That's not good - and Gmsh encourages people to use Physical Entities, so that they know what they need from a model and which ID they assign to it (so that later on, they would use it to distinguish parts corresponding to different materials, or boundary conditions). Gmsh accepts 0 as a user-defined physical IDs though, but there are some issues when such mesh is visualized in Gmsh, so a healthy grid would only have positive physical IDs. That's why I think MFEM shouldn't silently read the elements saved not by a user, or with bad IDs (which is 0), and assign 1 instead 0, or reassign all of them.

Thanks,
Mikhail

**Comment by tzanio** [comment]
Thanks Mikhail,

Internally we made a few adjustments, including throwing an error for non-positive attributes. These will be in the next release that should be coming out in the next few months.

Great pull request!

Tzanio
